File: Labs/Lab 4/TreeNode.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
    
    def invertTree(self, root):
        if root is None:
            return None
        
        root.left, root.right = root.right, root.left

        if root.left:
            self.invertTree(root.left)
        if root.right:
            self.invertTree(root.right)
        
        return root

# # Test Case
# tree = TreeNode(1)
# tree.left = TreeNode(2)
# tree.right = TreeNode(3)
# tree.left.left = TreeNode(4)
# tree.left.right = TreeNode(5)

# print(tree.val)
# print(tree.left.val)
# print(tree.right.val)
# print(tree.left.left.val)
# print(tree.left.right.val)

# tree.invertTree(tree)

# print()
# print(tree.val)
# print(tree.left.val)
# print(tree.right.val)
# print(tree.right.left.val)
# print(tree.right.right.val)

File: Labs/Lab4/TreeNode.java
package Labs.Lab4;

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(int val) {this.val = val;}
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public TreeNode invertTree(TreeNode root) {
        if(root == null){
            return null;
        }

        TreeNode temp = root.left;
        root.left = root.right;
        root.right = temp;

        invertTree(root.left);
        invertTree(root.right);

        return root;
    }

    // public static void main(String[] args) {
    //     TreeNode tree = new TreeNode(1);
    //     tree.left = new TreeNode(2);
    //     tree.right = new TreeNode(3);
    //     tree.left.left = new TreeNode(4);
    //     tree.left.right = new TreeNode(5);

    //     System.out.println(tree.val);
    //     System.out.println(tree.left.val);
    //     System.out.println(tree.right.val);
    //     System.out.println(tree.left.left.val);
    //     System.out.println(tree.left.right.val);

    //     tree.invertTree(tree);

    //     System.out.println(tree.val);
    //     System.out.println(tree.left.val);
    //     System.out.println(tree.right.val);
    //     System.out.println(tree.right.left.val);
    //     System.out.println(tree.right.right.val);
    // }
}